feedback/ngffeedback.cpp

```diff
@@ -0,0 +1,318 @@
+}
+
+NGFFeedback::~NGFFeedback()
+{
+    qCDebug(ngflc) << "Deinitializing plugin";
+}
+
+void NGFFeedback::failed(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCWarning(ngflc) << "Effect failed, id" << id;
+        ActiveEffect effect = m_activeEffects.take(id);
+        if (effect.customEffect)
+            m_customEffects.remove(effect.customEffect);
+    }
+}
+
+void NGFFeedback::completed(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCDebug(ngflc) << "Effect completed, id" << id;
+        ActiveEffect effect = m_activeEffects.take(id);
+        if (effect.customEffect)
+            m_customEffects.remove(effect.customEffect);
+    }
+}
+
+void NGFFeedback::playing(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCDebug(ngflc) << "Effect playing, id" << id;
+        m_activeEffects[id].state = QFeedbackEffect::Running;
+    }
+}
+
+void NGFFeedback::paused(quint32 id)
+{
+    if (m_activeEffects.contains(id)) {
+        qCDebug(ngflc) << "Effect paused, id" << id;
+        m_activeEffects[id].state = QFeedbackEffect::Paused;
+    }
+}
+
+bool NGFFeedback::play(QFeedbackEffect::Effect effect)
+{
+    quint32 id;
+    switch (effect) {
+    case QFeedbackEffect::Press:
+    case QFeedbackEffect::Release:
+    case QFeedbackEffect::PressWeak:
+    case QFeedbackEffect::ReleaseWeak:
+    case QFeedbackEffect::PressStrong:
+    case QFeedbackEffect::ReleaseStrong:
+    case QFeedbackEffect::DragStart:
+    case QFeedbackEffect::DragDropInZone:
+    case QFeedbackEffect::DragDropOutOfZone:
+    case QFeedbackEffect::DragCrossBoundary:
+        qCDebug(ngflc) << "Playing effect #" << effect << "(" << m_effects[effect] << ")";
+        id = m_client.play(m_effects[effect]);
+        if (!id)
+            qCWarning(ngflc) << "Could not play effect";
+        else
+            m_activeEffects.insert(id, { QFeedbackEffect::Running, nullptr });
+        return true;
+    case QFeedbackEffect::Appear:
+    case QFeedbackEffect::Disappear:
+    case QFeedbackEffect::Move:
+        qCDebug(ngflc) << "Unsupported effect #" << effect;
+        break;
+    default:
+        qCDebug(ngflc) << "Unknown or undefined effect #" << effect;
+        break;
+    }
+
+    return false;
+}
+
+QFeedbackInterface::PluginPriority NGFFeedback::pluginPriority()
+{
+    return QFeedbackInterface::PluginLowPriority;
+}
+
+QList<QFeedbackActuator*> NGFFeedback::actuators()
+{
+    return QList<QFeedbackActuator*>() << m_actuator;
+}
+
+void NGFFeedback::setActuatorProperty(const QFeedbackActuator &, ActuatorProperty prop, const QVariant &value)
+{
+    if (prop == ActuatorProperty::Enabled) {
+        bool old = m_actuatorEnabled;
+        m_actuatorEnabled = value.toBool();
+        if (old != m_actuatorEnabled && !m_actuatorEnabled) {
+            /* Stop all effects */
+            for (auto it = m_activeEffects.begin(); it != m_activeEffects.end(); ++it) { // TODO: Would it make sense to iterate backwards?
```

Comment:
So, what I mean here is that if ngfd was queueing these, would it be better to start dropping events from the end of the queue so that it could perhaps start playing the next event but it would not have to stop and continue to next all the time.

Reply:
Ehm... There is one problem with my thinking: QHash is iterated in non-deterministic order.

Reply:
Changed the underlying implementation to use QVector. Worse time-complexity but assuming a small number of items it shouldn't make a difference (why would you play a dozen effects at the same time?).